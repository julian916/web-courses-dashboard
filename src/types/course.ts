export type RateType = {
    stars: number
    amount: number
}

export enum LanguageEnum {
    EN = 'English',
    ES = 'Spanish',
    CH = 'Chinese',
    FR = 'French',
    PR = 'Portuguese',
    RU = 'Russian'
}

// ex: ['EN', 'ES', 'CH', 'FR', ...]
export const AvailableLanguage = Object.keys(LanguageEnum);

export type TeacherType = {
    name: string
    title: string
    profileImageUrl: string
}

export type LectureType = {
    name: string
    durationInSeconds: number
    description: string
}

export type ContentType = {
    lecturesAmount: number
    lecturesMinutes: number
    lectures: Array<LectureType>
}

export enum StatusEnum {
    ACTIVE = 'ACTIVE',
    COMPLETED = 'COMPLETED',
    NOT_STARTED = 'NOT_STARTED'
}

export enum LevelEnum {
    ALL = 'ALL',
    BEGINNER = 'BEGINNER',
    INTERMEDIATE = 'INTERMEDIATE',
    ADVANCED = 'ADVANCED'
}

export type CourseType = {
    id: number
    name: string
    description: string
    fullDescription: string
    rate: RateType
    level: LevelEnum
    language: LanguageEnum
    courseImageUrl: string
    introMediaUrl: string
    teacher: TeacherType
    content: ContentType
    status: StatusEnum
}
