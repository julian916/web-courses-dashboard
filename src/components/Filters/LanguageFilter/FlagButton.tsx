import React from 'react';

import { LanguageEnum } from "../../../types/course";
import styled from "styled-components";

type Props = {
  lang: LanguageEnum,
  onClick: (LanguageEnum) => void
  active: boolean
}

const StyledButton = styled.button`
  color: #000;
  background: transparent;
  padding: 8px 10px;
  border: 1px solid #dad8d8;
  border-radius: 10px;
  
  ${(props) => props.active ? `
    border: 1px solid black;
    text-shadow: 0.5px 0.5px;
  ` : ''}
`;

const StyledIcon = styled.img`
  width: 20px;
  height: 20px;
  border-radius: 50%;
`;

const StyledSpan = styled.span`
  margin: 0px 10px;
  font-size: 15px;
`;

const FlagButton = ({ lang, onClick, active } : Props) => {

  return (
    <StyledButton onClick={() => onClick(lang)} active={active}>
      <StyledIcon src={`flags/flag-${LanguageEnum[lang]}.svg`} alt={lang}/>
      <StyledSpan>
        {lang}
      </StyledSpan>
    </StyledButton>
  );
};

export default FlagButton;
