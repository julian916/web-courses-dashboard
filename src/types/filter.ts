import { LanguageEnum, StatusEnum } from "./course";

export type FilterType = {
    text: string | null
    status: StatusEnum | null
    language: LanguageEnum | null
}
