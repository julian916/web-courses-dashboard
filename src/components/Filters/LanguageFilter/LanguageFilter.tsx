import React from 'react';
import FlagButton from './FlagButton';
import styled from 'styled-components';

const Wrapper = styled.div`
  width: 85%;
  display: flex;
  justify-content: space-between;
  height: 45px;
`;

type Props = {
  onChange: (FiltersType) => void,
  language?: string | null
}

// TODO: Move this to a config file
const DISPLAYED_LANGUAGES = ['EN', 'ES', 'FR', 'CH'];

const LanguageFilter = ({ onChange, language } : Props) => {
  const handleSelect = (languageSelected) => {
    const newValue = languageSelected === language ? null : languageSelected;
    onChange({ language: newValue });
  };
  return (
    <Wrapper>
      {DISPLAYED_LANGUAGES.map((lang) => (
        <FlagButton
          lang={lang}
          onClick={handleSelect}
          active={lang === language}
          key={lang}
          id={lang}
        />
      ))}
    </Wrapper>
  );
};

export default LanguageFilter;
