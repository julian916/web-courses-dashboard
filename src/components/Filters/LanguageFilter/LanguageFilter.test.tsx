import React from 'react';
import { mount } from 'enzyme';
import LanguageFilter from "./LanguageFilter";
import FlagButton from "./FlagButton";
import { LanguageEnum } from "../../../types/course";

describe('<LanguageFilter />', () => {
  const onChange = jest.fn();
  const wrapper = mount(<LanguageFilter onChange={onChange} />);

  it('should render correctly initial state', () => {
    expect(wrapper).toMatchSnapshot();

    const buttons = wrapper.find(FlagButton);
    buttons.forEach(button => {
      expect(button.props().active).toBe(false);
    });
  });

  it('should show active first FlagButton after click', () => {
    const firstButton = wrapper.find(FlagButton).at(0);
    firstButton.simulate('click');
    wrapper.update();

    // Validate to have active only in clicked button
    expect(wrapper.find(FlagButton).at(0).props().active).toBe(true);
    expect(wrapper.find(FlagButton).at(1).props().active).toBe(false);

    // Validate that onChange was called with key value of enum
    expect(onChange).toHaveBeenCalledWith(LanguageEnum[firstButton.props().lang]);
  });

});
