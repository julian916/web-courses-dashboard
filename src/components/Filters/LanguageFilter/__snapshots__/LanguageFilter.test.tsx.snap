// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LanguageFilter /> should render correctly initial state 1`] = `ReactWrapper {}`;
